### PR TITLE
locale_maketext_test

### PR DESCRIPTION
**Pull request by raunakkathuria**

### CODE REVIEW

File: Makefile.PL [hunk 1]
```diff
@@ -0,0 +1,52 @@
+use 5.014;
+use strict;
+use warnings;
+use ExtUtils::MakeMaker;
+
+WriteMakefile(
+    NAME               => 'Locale::Maketext::Test',
+    AUTHOR             => q{Binary.com <binary@cpan.org>},
+    VERSION_FROM       => 'lib/Locale/Maketext/Test.pm',
+    ABSTRACT           => 'Locale::Maketext::Test supports testing of po files',
```
Comment (tom-binary) [inline Makefile.PL]: The name isn't needed here, since the abstract is usually displayed [next to the name](https://metacpan.org/author/BINARY) anyway - maybe just "test validity of .po translation files" would do?

File: lib/Locale/Maketext/Test.pm [hunk 2]
```diff
@@ -0,0 +1,565 @@
...
+=cut
+
+sub BUILD {
+    my $self = shift;
+
+    unless (scalar @{$self->languages}) {
+        my @lang = ();
+        if (opendir my $dh, $self->directory) {
+            while (readdir $dh) {
+                if (my ($x) = grep { $_ =~ /^(\w+)\.po$/ } $_) {
```
Comment (tom-binary) [inline lib/Locale/Maketext/Test.pm]: Since there's only a single item in the list:

``` perl
if(my ($name) = /^(\w+)\.po$/) { ... }
```

or maybe File::Basename to extract the name and skip anything that doesn't end in .po?

One issue with using readdir directly is that the filenames will be raw bytes, typically UTF-8 for Linux filesystems, so this might skip some files, e.g. ネコ.po. I'm not sure how common those are, and you can't just decode from UTF-8 since not all filesystems (particularly on Windows) use that...

File: lib/Locale/Maketext/Test.pm [hunk 3]
```diff
@@ -0,0 +1,565 @@
...
+
+    my (%header, @ids, $ln);
+    my $first = 1;
+
+    open my $f, '<:encoding(UTF-8)', $lang or die "Cannot open $lang: $!\n";
+    READ:
+    while (defined(my $l = _nextline($f))) {
+        if ($l =~ /^\s*msgid\s*"(.*)"/) {
+            my $line = $1;
+            $ln = $.;
```
Comment (tom-binary) [inline lib/Locale/Maketext/Test.pm:442]: Since _nextline maintains a stack, is $. going to be accurate?

File: lib/Locale/Maketext/Test.pm [hunk 4]
```diff
@@ -0,0 +1,565 @@
...
+        my $hnd = Locale::Maketext::ManyPluralForms->get_handle($lg);
+        $hnd->plural(1, 'test %d');
+
+        my $plural_sub = $hnd->{_plural};
+
+        my $nplurals = 2;    # default
+        $nplurals = $1 if $po->{header}->{'plural-forms'} =~ /\bnplurals=(\d+);/;
+        my @plural;
+
+        for (my ($i, $j) = (0, $nplurals); $i < 10000 && $j > 0; $i++) {
```
Comment (basiliscos) [inline lib/Locale/Maketext/Test.pm:224]: what is `10000`?

File: lib/Locale/Maketext/Test.pm [hunk 5]
```diff
@@ -0,0 +1,565 @@
...
+            $ln = $test->[3];
+            my $i     = 0;
+            my $j     = 0;
+            my @param = map {
+                $j++;
+                push @{$self->_status->{warnings}->{$lg}}, $self->_format_message($ln, "unused parameter \%$j") if (not defined $_ and $self->debug);
+                defined $_ && $_ eq 'text' ? 'text' . $i++ : 1;
+            } @{$test->[1]};
+            try {
+                local $SIG{__WARN__} = sub { die $_[0] };
```
Comment (basiliscos) [inline lib/Locale/Maketext/Test.pm:307]: may be move this out of the `for` cycle?

File: lib/Locale/Maketext/Test.pm (hunk 2 above)
Comment (evostrov) [inline lib/Locale/Maketext/Test.pm]: @raunakkathuria 
Why do you grep by single string?
I think you can do

``` perl
if (my ($x) = ($_ =~ /^(\w+)\.po$/)) {
  push @lang, $x;
}
```

Comment (tom-binary) [inline lib/Locale/Maketext/Test.pm]: or even:

``` perl
push @lang, /^(\w+)\.po$/;
```

since that's an empty list without a match (that's what makes the `if` clause work in the first place). If you find the explicit `if()` more readable I'm fine with that.

File: t/locales/ru.po [hunk 6]
```diff
@@ -0,0 +1,27 @@
+# SOME DESCRIPTIVE TITLE.
+# Copyright (C) YEAR THE PACKAGE'S COPYRIGHT HOLDER
+# This file is distributed under the same license as the PACKAGE package.
+# FIRST AUTHOR <EMAIL@ADDRESS>, YEAR.
+msgid ""
+msgstr ""
```
Comment (sasha-murzin) [inline t/locales/ru.po:6]: may be just remove this?

File: lib/Locale/Maketext/Test.pm (hunk 3 above)
Comment (torsten-deriv) [inline lib/Locale/Maketext/Test.pm:442]: I think it is. But I think better to either add a comment why or to stash `$.` together with the line.

Here is why I think it is correct. A continuation line is marked with a leading space. Such lines are never un-read. When the program finds the next non-continuation, it un-reads it. At this point `$.` points at the un-read line. The next step is then to fetch the line again, this time from the stack. If it is a new beginning, `$.` is read. But it still has the accurate value. If it is not a new beginning, comment or empty line or junk, then the next line is read from the file which means `$.` is updated.

So, basically it boils down to yes it is accurate because the number of un-read lines is <=1.

File: lib/Locale/Maketext/Test.pm (hunk 2 above)
Comment (raunakkathuria) [inline lib/Locale/Maketext/Test.pm]: @evostrov thats my bad, earlier it was array so was using grep but removed array and forgot to remove grep

File: lib/Locale/Maketext/Test.pm (hunk 4 above)
Comment (raunakkathuria) [inline lib/Locale/Maketext/Test.pm:224]: @basiliscos its just a random number to capture all plurals, we can any lower number as well

Comment (torsten-deriv) [inline lib/Locale/Maketext/Test.pm:224]: That is just an arbitrary _high_ number. The assumption is that for every natural language all plural forms are met at least once for quantities from 0 to 10000. Not sure if that is true.

Note that the loop only runs until each expected plural form is met once.
